api/wft-complete-metadata.md

```diff
@@ -0,0 +1,126 @@
+# Workflow Task Completion Metadata
+
+We have a need to attach various bits of metadata to workflow task completions. There are a number
+of usecases in which the SDK would like to save some information for itself, or tell server about
+something, but this information does not influence replay or the reconstitution of workfow state in
+any way. Thus, it shouldn't go in a marker or some new kind of history event, because this would
+violate the (useful) idea that history events are primarily concerned with recreating workflow
+state.
+
+## Use cases
+
+### Reporting currently-being-awaited commands & facts about them
+
+Right now, besides a query, there's no simple way for a user to understand what their workflow is
+actively waiting on. Solving this would make it much easier for users to have insight into why
+their workflow is or isn't making progress, and could be nicely exposed in the UI.
+
+In particular, understanding what's going on while your workflow is currently running local
+activities and performing WFT heartbeats is very opaque. This would make that much more clear.
+
+Imagine being able to get information like this (and do useful things with it in the UI):
+
+```
+waiting_on: [
+    { command_type: "local_activity", activity_name: "la_func", id: 1, retries: 0, location: "my_file.py:23" },
+    { command_type: "local_activity", activity_name: "la_func", id: 2, retries: 5, location: "my_file.py:23" },
+    { command_type: "timer", id: 3, duration: "5s", location: "my_file.py:42" },
+]
+```
+
+### SDK internal versioning
+
+We would like to be able to make SDK changes which affect how history is written, which means
+flagging our behavior based on what version of the SDK the history was produced with. Note: *not*
+relating to concepts of versioning a *user's* code, but rather the SDK itself.
+
+This internal version could easily slot into this metadata.
+
+Note: This could potentially go in `temporal.api.taskqueue.v1.VersionId` too, but that's more
+intended to relate to user code.
+
+### Dynamic config adjustments
+
+Allow the worker to set configurable things which don't directly impact determinism or workflow
+state. For example, changing the WFT timeout for the next task.
+
+### Various debugging / analysis information
+
+There's a whole world of little bits of info that could be useful to enable selectively to help
+users debug their workflows, help us learn about performance information, etc.
+
+Some examples:
+* Emit the OTel span id that was associated with the WFT - now the user can go easily look up any
+  logs/events that were emitted as part of that WFT.
+* Metrics work great for all sorts of latency measurements, but what if you wanted to know about
+  how long a _specific_ WFT took? You could emit that.
+* Interceptors could pipe whatever arbitrary data the user wants into there, for use cases where
```

Comment:
Spans already allow you to store arbitrary key/value pairs. Why can't we just use that?